src/bigquery.ts

```diff
@@ -2391,9 +2371,228 @@ promisifyAll(BigQuery, {
+ * The lower and upper bound for the range are optional.
+ * The lower bound is inclusive and the upper bound is exclusive.
+ * See https://cloud.google.com/bigquery/docs/reference/standard-sql/lexical#range_literals
+ */
+export class BigQueryRange {
+  elementType?: string;
+  start?: BigQueryTimestamp | BigQueryDate | BigQueryDatetime;
+  end?: BigQueryTimestamp | BigQueryDate | BigQueryDatetime;
+  constructor(value: string | BigQueryRangeOptions, elementType?: string) {
+    if (typeof value === 'string') {
+      if (!elementType) {
+        throw new Error(
+          'invalid RANGE. Element type required when using RANGE literal string.'
+        );
+      }
+      let cleanedValue = value;
+      if (cleanedValue.startsWith('[') || cleanedValue.startsWith('(')) {
+        cleanedValue = cleanedValue.substring(1);
+      }
+      if (cleanedValue.endsWith(')') || cleanedValue.endsWith(']')) {
+        cleanedValue = cleanedValue.substring(0, cleanedValue.length - 1);
+      }
+      const parts = cleanedValue.split(',');
+      if (parts.length !== 2) {
+        throw new Error(
+          'invalid RANGE. See RANGE literal format docs for more information.'
+        );
+      }
+
+      const [start, end] = parts.map((s: string) => s.trim());
+      this.start = this.convertElement_(start, elementType);
+      this.end = this.convertElement_(end, elementType);
+      this.elementType = elementType;
+    } else {
+      const {start, end} = value;
+      if (start && end) {
+        if (typeof start !== typeof end) {
+          throw Error(
+            'upper and lower bound on a RANGE should be of the same type.'
+          );
+        }
+      }
+      const inferredType =
+        {
+          BigQueryDate: 'DATE',
+          BigQueryDatetime: 'DATETIME',
+          BigQueryTimestamp: 'TIMESTAMP',
+        }[(start || end || Object).constructor.name] || elementType;
+      this.start = this.convertElement_(start, inferredType);
+      this.end = this.convertElement_(end, inferredType);
+      this.elementType = inferredType;
+    }
+  }
+
+  public get literalValue() {
+    return `[${this.start ? this.start.value : 'UNBOUNDED'}, ${this.end ? this.end.value : 'UNBOUNDED'})`;
+  }
+
+  public get value() {
+    return {
+      start: this.start ? this.start.value : 'UNBOUNDED',
+      end: this.end ? this.end.value : 'UNBOUNDED',
+    };
+  }
+
+  static fromSchemaValue_(value: string, elementType: string): BigQueryRange {
+    let cleanedValue = value;
+    if (cleanedValue.startsWith('[') || cleanedValue.startsWith('(')) {
+      cleanedValue = cleanedValue.substring(1);
+    }
+    if (cleanedValue.endsWith(')') || cleanedValue.endsWith(']')) {
+      cleanedValue = cleanedValue.substring(0, cleanedValue.length - 1);
+    }
+    const parts = cleanedValue.split(',');
+    if (parts.length !== 2) {
+      throw new Error(
+        'invalid RANGE. See RANGE literal format docs for more information.'
```

Comment:
Well, so this bit is mildly interesting.  The API representation `[val1, val2)` is a subset of the actual literal format which also includes the `RANGE<TYPE> ` prefix string to make it unambiguous.  So, should we handle both the literal and API format?

Reply:
I added methods to return the Literal format and API format now. But the Range class parses the API representation.